core/distributed/index_map.cpp

```diff
@@ -0,0 +1,187 @@
+{
+    return exec_;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+size_type index_map<LocalIndexType, GlobalIndexType>::get_global_size() const
+{
+    return partition_ ? partition_->get_size() : 0;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+array<LocalIndexType> index_map<LocalIndexType, GlobalIndexType>::map_to_local(
+    const array<GlobalIndexType>& global_ids, index_space index_space_v) const
+{
+    array<LocalIndexType> local_ids(exec_);
+
+    exec_->run(index_map_kernels::make_map_to_local(
+        partition_.get(), remote_target_ids_, to_device(remote_global_idxs_),
+        rank_, global_ids, index_space_v, local_ids));
+
+    return local_ids;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+index_map<LocalIndexType, GlobalIndexType>::index_map(
+    std::shared_ptr<const Executor> exec,
+    std::shared_ptr<const partition_type> partition, comm_index_type rank,
+    const array<GlobalIndexType>& recv_connections)
+    : exec_(std::move(exec)),
+      partition_(std::move(partition)),
+      rank_(rank),
+      remote_target_ids_(exec_),
+      remote_local_idxs_(exec_),
+      remote_global_idxs_(exec_)
+{
+    array<LocalIndexType> flat_remote_local_idxs(exec_);
+    array<GlobalIndexType> flat_remote_global_idxs(exec_);
+    array<int64> remote_sizes(exec_);
+    exec_->run(index_map_kernels::make_build_mapping(
+        partition_.get(), recv_connections, remote_target_ids_,
+        flat_remote_local_idxs, flat_remote_global_idxs, remote_sizes));
+    remote_local_idxs_ = segmented_array<LocalIndexType>::create_from_sizes(
+        std::move(flat_remote_local_idxs), remote_sizes);
+    remote_global_idxs_ = segmented_array<GlobalIndexType>::create_from_sizes(
+        std::move(flat_remote_global_idxs), remote_sizes);
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+index_map<LocalIndexType, GlobalIndexType>::index_map(
+    std::shared_ptr<const Executor> exec)
+    : exec_(exec),
+      partition_(nullptr),
+      remote_target_ids_(exec),
+      remote_local_idxs_(exec),
+      remote_global_idxs_(exec)
+{}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+index_map<LocalIndexType, GlobalIndexType>&
+index_map<LocalIndexType, GlobalIndexType>::operator=(const index_map& other)
+{
+    if (this != &other) {
+        partition_ = other.partition_;
+        rank_ = other.rank_;
+        remote_target_ids_ = other.remote_target_ids_;
+        remote_local_idxs_ = other.remote_local_idxs_;
+        remote_global_idxs_ = other.remote_global_idxs_;
+    }
+    return *this;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+index_map<LocalIndexType, GlobalIndexType>&
+index_map<LocalIndexType, GlobalIndexType>::operator=(index_map&& other)
+{
+    if (this != &other) {
+        partition_ = std::move(other.partition_);
+        rank_ = other.rank_;
```

Comment:
nit: should the moved-from state preserve the original rank?

Reply:
I think that is sensible. We don't really have an invalid rank value that I could put here.